File: wildlife-tracking/detection/tracker.py
"""
Wildlife Tracking Engine
Implements ByteTrack with Centroid-IoU fallback tracker across camera trap frames.
"""
import math
import time
from typing import List, Dict, Any, Tuple

class FallbackIoUTracker:
    """
    Centroid-Distance & IoU-Matching Tracker.
    Disclosed fallback when ByteTrack/supervision packages are not installed.
    Persists tracking IDs and counts unique wildlife individuals across frames.
    """
    def __init__(self, iou_threshold: float = 0.25, max_age_seconds: float = 8.0):
        self.next_track_id = 1
        self.active_tracks: List[Dict[str, Any]] = []
        self.unique_track_ids = set()
        self.iou_threshold = iou_threshold
        self.max_age_seconds = max_age_seconds

    def update(self, detections: List[Dict[str, Any]]) -> List[Dict[str, Any]]:
        now = time.time()
        # Remove tracks older than max_age_seconds
        self.active_tracks = [t for t in self.active_tracks if (now - t["last_seen"]) < self.max_age_seconds]

        matched_track_indices = set()
        results = []

        for det in detections:
            bbox = det.get("bbox", [0, 0, 1, 1])
            class_idx = det.get("class_index", -1)
            best_match_idx = -1
            best_score = -1.0

            for idx, track in enumerate(self.active_tracks):
                if idx in matched_track_indices:
                    continue
                iou = self.calculate_iou(bbox, track["bbox"])
                centroid_dist = self.calculate_centroid_dist(bbox, track["bbox"])
                is_same_class = track["class_index"] == class_idx
                score = (iou * 0.7) + (0.3 if is_same_class else 0.0) - (centroid_dist * 0.2)

                if iou >= self.iou_threshold or (centroid_dist < 0.25 and is_same_class):
                    if score > best_score:
                        best_score = score
                        best_match_idx = idx

            if best_match_idx >= 0:
                matched_track_indices.add(best_match_idx)
                track = self.active_tracks[best_match_idx]
                track["bbox"] = bbox
                track["last_seen"] = now
                track["hits"] += 1
                assigned_id = track["track_id"]
            else:
                assigned_id = self.next_track_id
                self.next_track_id += 1
                self.active_tracks.append({
                    "track_id": assigned_id,
                    "class_index": class_idx,
                    "animal_name": det.get("animal_name", ""),
                    "bbox": bbox,
                    "last_seen": now,
                    "hits": 1
                })

            self.unique_track_ids.add(assigned_id)
            res = dict(det)
            res["tracking_id"] = assigned_id
            results.append(res)

        return results

    def get_unique_count(self) -> int:
        return len(self.unique_track_ids)

    def get_active_count(self) -> int:
        now = time.time()
        return len([t for t in self.active_tracks if (now - t["last_seen"]) < self.max_age_seconds])

    def reset(self):
        self.next_track_id = 1
        self.active_tracks.clear()
        self.unique_track_ids.clear()

    @staticmethod
    def calculate_iou(box_a, box_b):
        y_a1, x_a1, y_a2, x_a2 = box_a
        y_b1, x_b1, y_b2, x_b2 = box_b
        y_inter1 = max(y_a1, y_b1)
        x_inter1 = max(x_a1, x_b1)
        y_inter2 = min(y_a2, y_b2)
        x_inter2 = min(x_a2, x_b2)
        inter_area = max(0, y_inter2 - y_inter1) * max(0, x_inter2 - x_inter1)
        area_a = max(0, y_a2 - y_a1) * max(0, x_a2 - x_a1)
        area_b = max(0, y_b2 - y_b1) * max(0, x_b2 - x_b1)
        union_area = area_a + area_b - inter_area
        return inter_area / union_area if union_area > 0 else 0.0

    @staticmethod
    def calculate_centroid_dist(box_a, box_b):
        c_ay = (box_a[0] + box_a[2]) / 2.0
        c_ax = (box_a[1] + box_a[3]) / 2.0
        c_by = (box_b[0] + box_b[2]) / 2.0
        c_bx = (box_b[1] + box_b[3]) / 2.0
        return math.sqrt((c_ay - c_by)**2 + (c_ax - c_bx)**2)

def get_tracker():
    try:
        import supervision as sv
        # ByteTrack implementation
        return sv.ByteTrack()
    except ImportError:
        # Transparent Fallback
        return FallbackIoUTracker()

File: server/tracker.ts
export interface Track {
  track_id: number;
  class_index: number;
  animal_name: string;
  bbox: [number, number, number, number]; // [ymin, xmin, ymax, xmax]
  last_seen: number;
  hits: number;
}

export interface BoundingBox {
  ymin: number;
  xmin: number;
  ymax: number;
  xmax: number;
}

export class WildlifeTracker {
  private nextTrackId: number = 1;
  private activeTracks: Track[] = [];
  private totalUniqueTracks: Set<number> = new Set();
  private maxAgeMs: number = 8000; // 8 seconds before track expires
  private iouThreshold: number = 0.25;

  public update(
    detections: Array<{
      class_index: number;
      animal_name: string;
      confidence: number;
      bbox: [number, number, number, number];
    }>
  ): Array<{
    class_index: number;
    animal_name: string;
    confidence: number;
    bbox: [number, number, number, number];
    tracking_id: number;
  }> {
    const now = Date.now();
    // Prune stale tracks
    this.activeTracks = this.activeTracks.filter(t => now - t.last_seen < this.maxAgeMs);

    const results: Array<{
      class_index: number;
      animal_name: string;
      confidence: number;
      bbox: [number, number, number, number];
      tracking_id: number;
    }> = [];

    const matchedTrackIndices = new Set<number>();

    for (const det of detections) {
      let bestMatchIdx = -1;
      let bestScore = -1;

      for (let i = 0; i < this.activeTracks.length; i++) {
        if (matchedTrackIndices.has(i)) continue;
        const track = this.activeTracks[i];
        
        // Calculate IoU & Centroid Distance
        const iou = calculateIoU(det.bbox, track.bbox);
        const centroidDist = calculateCentroidDist(det.bbox, track.bbox);

        // Same class or high overlap
        const isSameClass = track.class_index === det.class_index;
        const score = (iou * 0.7) + (isSameClass ? 0.3 : 0) - (centroidDist * 0.2);

        if (iou >= this.iouThreshold || (centroidDist < 0.25 && isSameClass)) {
          if (score > bestScore) {
            bestScore = score;
            bestMatchIdx = i;
          }
        }
      }

      let assignedTrackId: number;
      if (bestMatchIdx >= 0) {
        matchedTrackIndices.add(bestMatchIdx);
        const track = this.activeTracks[bestMatchIdx];
        track.bbox = det.bbox;
        track.last_seen = now;
        track.hits += 1;
        assignedTrackId = track.track_id;
      } else {
        // Create new track
        assignedTrackId = this.nextTrackId++;
        this.activeTracks.push({
          track_id: assignedTrackId,
          class_index: det.class_index,
          animal_name: det.animal_name,
          bbox: det.bbox,
          last_seen: now,
          hits: 1
        });
      }

      this.totalUniqueTracks.add(assignedTrackId);

      results.push({
        ...det,
        tracking_id: assignedTrackId
      });
    }

    return results;
  }

  public getUniqueCount(): number {
    return this.totalUniqueTracks.size;
  }

  public getActiveCount(): number {
    const now = Date.now();
    return this.activeTracks.filter(t => now - t.last_seen < this.maxAgeMs).length;
  }

  public reset(): void {
    this.nextTrackId = 1;
    this.activeTracks = [];
    this.totalUniqueTracks.clear();
  }
}

function calculateIoU(boxA: [number, number, number, number], boxB: [number, number, number, number]): number {
  const [yA1, xA1, yA2, xA2] = boxA;
  const [yB1, xB1, yB2, xB2] = boxB;

  const yInter1 = Math.max(yA1, yB1);
  const xInter1 = Math.max(xA1, xB1);
  const yInter2 = Math.min(yA2, yB2);
  const xInter2 = Math.min(xA2, xB2);

  const interArea = Math.max(0, yInter2 - yInter1) * Math.max(0, xInter2 - xInter1);
  const areaA = Math.max(0, yA2 - yA1) * Math.max(0, xA2 - xA1);
  const areaB = Math.max(0, yB2 - yB1) * Math.max(0, xB2 - xB1);
  const unionArea = areaA + areaB - interArea;

  return unionArea > 0 ? interArea / unionArea : 0;
}

function calculateCentroidDist(boxA: [number, number, number, number], boxB: [number, number, number, number]): number {
  const cAy = (boxA[0] + boxA[2]) / 2;
  const cAx = (boxA[1] + boxA[3]) / 2;
  const cBy = (boxB[0] + boxB[2]) / 2;
  const cBx = (boxB[1] + boxB[3]) / 2;

  const dy = cAy - cBy;
  const dx = cAx - cBx;
  return Math.sqrt(dy * dy + dx * dx);
}

export const globalWildlifeTracker = new WildlifeTracker();
